File: algorithms/code/hackerrank/staircase/staircase.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/staircase

import unittest


def staircase(n):
    result = []
    for index in range(n):
        line = ''
        for i in range(n):
            line += ' ' if i < n - 1 - index else '#'
        result.append(line)
    return result


class TestCode(unittest.TestCase):
    def test_example(self):
        expected = ['     #', '    ##', '   ###', '  ####', ' #####', '######']
        self.assertEqual(expected, staircase(6))

File: algorithms/code/hackerrank/staircase/staircase.rs
// https://www.hackerrank.com/challenges/staircase

pub fn staircase(n: i32) -> Vec<String> {
    let mut result = Vec::new();
    for index in 0..n {
        let mut line = String::from("");
        for i in 0..n {
            if i < n - 1 - index {
                line.push(' ');
            } else {
                line.push('#');
            }
        }
        result.push(line);
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_example() {
        let expected = vec![
            String::from("     #"),
            String::from("    ##"),
            String::from("   ###"),
            String::from("  ####"),
            String::from(" #####"),
            String::from("######"),
        ];
        assert_eq!(expected, staircase(6));
    }
}
